### PR TITLE
Make GraphQL AST emission lazy

### PR DESCRIPTION
**Pull request by FelipeCoimbra**
TypeScript node collection in ts2gql is lazy, meaning it only transverses nodes of the `.ts` file which are somehow referenced from root `interface Schema` declaration or its children while building the GraphQL AST dynamically.

This PR proposes the same behavior for emitting the transpiled GraphQL AST. Besides being more performatic and making the generated schema more readable, this allows the user to truly benefit from the `DefinitionAlias` feature. With the lazy behavior, the transpiler would only emit the original definition of some GraphQL Type if it's directly referenced. For example, currently the declarations:
```ts
interface A {
    sa:boolean;
}
type NullableA = A | null;
interface Query {
    test:NullableA;
}
/** @graphql Schema */
export interface Schema {
    query:Query;
}
```
Generates
```gql
type A {
  sa: Boolean!
}
type NullableA {
  sa: Boolean!
}
type Query {
  test: NullableA
}
schema {
  query: Query
}
```
While with a lazy behavior it would only emit
```gql

type NullableA {
  sa: Boolean!
}
type Query {
  test: NullableA
}
schema {
  query: Query
}
```

### CODE REVIEW

File: src/Emitter.ts [hunk 1]
```diff
@@ -18,17 +20,32 @@ export default class Emitter {
...
   emitAll(stream:NodeJS.WritableStream) {
     stream.write('\n');
-    this.typeMap.forEach((node, name) => {
-      const content = this.emitTopLevelNode(node, name, stream);
-      if (content) {
-        stream.write(`${content}\n\n`);
-      }
-    });
+    const query = this.typeMap.get(this.root.query);
+    let mutation;
```
Comment (diegoximenes) [inline src/Emitter.ts]: nitpicking: you can use a ternary operator and use `const` instead of `let`.

File: src/Emitter.ts [hunk 2]
```diff
@@ -18,17 +20,32 @@ export default class Emitter {
...
+    if (this.root.mutation) {
```
Comment (diegoximenes) [inline src/Emitter.ts]: You don't check `this.root.query` to define `query`, why?

File: src/Emitter.ts [hunk 3]
```diff
@@ -18,17 +20,32 @@ export default class Emitter {
...
+    if (mutation) {
+      const mutationRootName = this._name(this.root.mutation!);
+      this.emitTopLevelNode(mutation, mutationRootName);
+    }
+    this.emissionQueue.forEach(emissionElem => stream.write(`${this.emissionMap.get(emissionElem)!}\n`));
     stream.write(`${this.emitSchema()}\n`);
   }
 
-  emitTopLevelNode(node:types.TypeDefinitionNode, name:types.SymbolName, stream:NodeJS.WritableStream):string {
+  emitTopLevelNode(node:types.TypeDefinitionNode, name:types.SymbolName) {
```
Comment (diegoximenes) [inline src/Emitter.ts]: nitpicking: `_emitTopLevelNode`. Same thing: `_emitSchema`

**Review by diegoximenes — APPROVED** [review]
Did some tests and it is working!
I am not familiar with the code though, therefore my review was not complete.

File: src/Emitter.ts (hunk 2 above)
Comment (FelipeCoimbra) [reply src/Emitter.ts]: this.root.query is garanteed to exist because GraphQL demands Query to always be defined, while Mutation is optional

**Pull request merged**